pyns/endpoints/base.py

```diff
@@ -42,3 +64,108 @@ def _base_path_(self):
     def _auto_methods_(self):
         """ HTTP methods to auto create in subordinate classes """
         pass
+
+
+def names_to_ids(func):
+    ''' Decorator which converts _name to _id by automatically looking up in API'''
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        api = pyns.Neuroscout()
+        # Sorting guarantees that 'dataset' is looked up before 'task' and 'predictor' to constrain search
+        for kw in sorted(kwargs):
+            if kw.endswith('_name'):
+                try:
+                    mod = getattr(api, kw.replace('_name', 's'))
+                except:
+                    raise ValueError("No API endpoint for {}".format(kw))
+
+                filter_args = {'name': kwargs[kw]}
+                if kw == 'task_name' and 'dataset_id' in kwargs:
+                    filter_args['dataset_id'] = kwargs['dataset_id']
+
+                if kw == 'predictor_name' and 'run_id' in kwargs: 
+                    filter_args['run_id'] = kwargs['run_id']
+                
+                res = mod.get(**filter_args)
+
+                if not res:
+                    raise ValueError("No {} found using provided arguments".format(kw))
+                if len(res) > 1:
+                    raise ValueError("Multiple {} found using provided arguments".format(kw))
+
+                kwargs[kw.replace('_name', '_id')] = res[0]['id']
+                kwargs.pop(kw)
+
+        return func(*args, **kwargs)
+    return wrapper
+
+
+def find_runs(func):
+    """ Decorator which finds runs for a given dataset and task names.
+    Assumes that downstream function accepts names instead of ids 
+    (i.e. has been decorated with names_to_ids)"""
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        api = pyns.Neuroscout()
+        FIELDS = ['dataset_name', 'task_name', 'subject', 'number', 'session']
+        search_args = {k: kwargs.pop(k) for k in FIELDS if k in kwargs}
+
+        if 'run_ids' not in kwargs and search_args:
+            kwargs['run_id'] = [r['id'] for r in api.runs.get(**search_args)]
+            if not kwargs['run_id']:
+                raise ValueError("No runs found using provided arguments")
+        elif 'run_ids' in kwargs and search_args:
+            raise ValueError(f"Run filter arguments {search_args} cannot be provided if run_ids are provided")
+        return func(*args, **kwargs)
+    return wrapper
+
+
+def _id_to_entities(df):
+    """ Given a column of ids, return a column of names, or in the case of
+    run_id, a column of dataset_name, task_name, and predictor_name """
+    api = pyns.Neuroscout()
+    for col in df.columns:
+        if col.endswith('_id'):
+            try:
+                endpoint = getattr(api, col.replace('_id', 's'))
+            except:
+                warnings.warn(f"No API endpoint for {col}, could not convert")
+                
+            if col == 'run_id':
+                ents = ['subject', 'session', 'number', 'acquisition']
```

Comment:
minor thing, this is not used - so probably remove

Reply:
Why are you saying this is not used?

This adds from `run_id` the columns `subject` `session` etc...

Reply:
I'm merging, but if you remember let me know

Reply:
the variable `ents` is not used, as far as I can see?